### PR TITLE
Nl solver

### PR DESCRIPTION
**Pull request by teseoch**
Missing bits
- [ ] unit tests
     - [x] quadratic
     - [x] Rosenbrock
     - [ ] a couple from https://en.wikipedia.org/wiki/Test_functions_for_optimization
     - [ ] box constraints
- [x] MMA
     - [x] MMA inside solver
- [x] scoped timers and timers
- [ ] stochastic grad descent
- [x] dense linear solvers
- [x] factory
- [ ] todos

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -133,26 +133,46 @@ endif()
 ################################################################################
 
 # Add an empty library and fill in the list of sources in `src/CMakeLists.txt`.
+add_library(polysolve_linear)
+add_library(polysolve::polysolve_linear ALIAS polysolve_linear)
```
[the remark this answers is not in the document]
Comment (teseoch) [reply CMakeLists.txt:137]: ok

File: CMakeLists.txt [hunk 2]
```diff
@@ -272,8 +294,6 @@ endif()
 # Compiler options
 ################################################################################
 
-# Use C++14
-target_compile_features(polysolve PUBLIC cxx_std_14)
```
[the remark this answers is not in the document]
Comment (teseoch) [reply CMakeLists.txt:276]: i have these on top
```
    set(CMAKE_CXX_STANDARD 17)
    set(CMAKE_CXX_STANDARD_REQUIRED ON)
    set(CMAKE_CXX_EXTENSIONS OFF)
```
is this not enough?

[the remark this answers is not in the document]
Comment (zfergus) [reply CMakeLists.txt:276]: Thats only if POLYSOLVE_TOPLEVEL_PROJECT

File: src/polysolve/nonlinear/Solver.cpp [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+            "Starting {} solve f₀={:g} ‖∇f₀‖={:g} "
+            "(stopping criteria: max_iters={:d} Δf={:g} ‖∇f‖={:g} ‖Δx‖={:g})",
+            name(), objFunc.value(x), this->m_current.gradNorm, this->m_stop.iterations,
+            this->m_stop.fDelta, this->m_stop.gradNorm, this->m_stop.xDelta);
+
+        update_solver_info(objFunc.value(x));
+
+        do
+        {
+            if (name() == "MMA")
```
Comment (teseoch) [inline src/polysolve/nonlinear/Solver.cpp:126]: this is bad

File: src/polysolve/nonlinear/Solver.cpp [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+            // ------------------------
+
+            // Compute a Δx to update the variable
+            if (!compute_update_direction(objFunc, x, grad, delta_x))
+            {
+                this->m_status = cppoptlib::Status::Continue;
+                continue;
+            }
+
+            if (name() != "MMA" && grad_norm != 0 && delta_x.dot(grad) >= 0)
```
Comment (teseoch) [inline src/polysolve/nonlinear/Solver.cpp:170]: this too

File: src/polysolve/nonlinear/Solver.cpp (hunk 3 above)
Comment (teseoch) [reply src/polysolve/nonlinear/Solver.cpp:174]: @Huangzizhou can this be refactored somehow? why is this necessary? same for the other one

**Pull request merged**